### ISSUE TITLE
Syntax for modifying records

### ISSUE THREAD

**Issue by damhiya** [ghc-proposals/ghc-proposals#453]
The `Record Dot Syntax` proposal [does not provides syntax for modifying records](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0282-record-dot-syntax.rst#should-a-syntax-be-provided-for-modification) because no appealing syntax has been suggested yet.

I thought out pretty good syntax for this, so I'll introduce it here.
```haskell
data Company = Company {name :: String, owner :: Person}
data Person = Person {name :: String, age :: Int}

getAged :: Company -> Company
getAged company = company { owner.age@age = age + 1 }
```

What do you think about it?

**Comment by adamgundry** [comment]
For context, `OverloadedRecordUpdate` (which is specified but not yet fully implemented) will allow one to write record update expressions such as
```hs
getAged :: Company -> Company
getAged company = company { owner.age = company.owner.age + 1 }
```
The original record dot syntax proposal suggested the following, but it was removed in the subsequent discussion:
```hs
getAged :: Company -> Company
getAged company = company { owner.age + 1 }
```

Personally I think we shouldn't introduce yet more syntactic variations here. Lenses already provide a powerful, compositional approach to writing record modifications. The main motivation for record dot syntax IMHO is to provide something simple, and if we add lots more syntax, that simplicity is lost.

**Comment by int-index** [comment]
I thought the main motivation for record dot syntax was scoping, so that in `r.x` the `x` is a symbol, not a top-level name. And I would appreciate syntax for
* modifications
* effectful modifications

To follow @damhiya’s suggestion of reusing `@` of as-patterns, I’d want to be able to write:

```
relocate :: (City -> City) -> Person -> Person
relocate f person =
  person { location.city@s = f s }

relocateIO :: (City -> IO City) -> Person -> IO Person
relocateIO f person =
  person { location.city@s <- f s }
```

**Comment by tomjaguarpaw** [comment]
One of the wonderful things about Haskell is that it composes well. A small core of simple components can be plugged together to build up more sophisticated behaviours in predictable ways. *If* this proposal makes the language as a whole simpler and more coherent then I am all for it. However I fear it is just piling on another layer of syntactic complexity. Eventually we will be crushed. Let's take a step back to catch our breath before considering even more [small syntax extensions](https://osa1.net/posts/2020-01-22-no-small-syntax-extensions.html).

**Comment by Ericson2314** [comment]
Yes please no more small record syntax proposals. I am already miffed with record dot IMO being "success at all costs" for side-stepping more subtle semantic issues like not working with fields which have polytypes to just "do the popular thing".

**Comment by danidiaz** [comment]
If a `SetField` typeclass existed, we could repurpose—with no extra changes—`OverloadedRecordDot` for setting, not only for getting:

    class SetField (n :: Symbol) r v | n r -> v where
        modifyField :: (v -> v) -> r -> r

    newtype Optick a b = Optick ((b -> b) -> a -> a)

    the :: Optick a a -- Control.Category.id ?
    the = Optick ($)  

    set :: Optick a b -> b -> a -> a
    set (Optick f) b = f (const b)

    modify :: Optick a b -> (b -> b) -> a -> a
    modify (Optick f) = f

    instance SetField n b c => HasField n (Optick a b) (Optick a c) where
        getField (Optick f) = Optick (\changec -> f (modifyField @n changec))


so we could write:

    getAged :: Company -> Company
    getAged company = modify the.owner.age (+1) company

Without having to enable `OverloadedRecordUpdate` and shadow normal type-changing record update syntax. 



[The full gist](https://gist.github.com/danidiaz/a945ff36ddccac0851fea2a4485df350#file-dotsetter5_setfield-hs)

**Comment by mbj** [comment]
> company { owner.age@age = age + 1 }

`@age` is not a type, its a field selection. But could easily be misread as a type application.

**Comment by int-index** [comment]
@mbj What’s your opinion on as-patterns then?

```
f x@(Just a) = (x, a)
```

**Comment by mbj** [comment]
@int-index They at least match the constructor, which is defined by the type. But than I could say the same for the field which is also defined by the type. So my argument is not as strong as I thought it would be.
